dev-tools/mage/integtest.go

```diff
@@ -242,19 +396,77 @@ func IsInIntegTestEnv() bool {
 	return found
 }
 
-func haveIntegTestEnvRequirements() error {
-	if err := HaveDockerCompose(); err != nil {
-		return err
+// IntegrationEnv represents the types of integrations environment the dir needs.
+type IntegrationEnv struct {
+	types map[string]string
+}
+
+// NewIntegrationEnvFromDir gets the integration environment for the dir.
+func NewIntegrationEnvFromDir(dir string) *IntegrationEnv {
+	cwd, err := os.Getwd()
+	if err != nil {
+		panic(err)
 	}
-	if err := HaveDocker(); err != nil {
-		return err
+	types := map[string]string{}
+	dockerFile := filepath.Join(cwd, dir, "docker-compose.yml")
+	if _, err := os.Stat(dockerFile); !os.IsNotExist(err) {
+		types[TestingEnvDocker] = dockerFile
+	}
+	kubeFile := filepath.Join(cwd, dir, "kubernetes.yml")
+	if _, err := os.Stat(kubeFile); !os.IsNotExist(err) {
+		types[TestingEnvKubernetes] = kubeFile
+	}
+	return &IntegrationEnv{types}
+}
+
+// HasType returns true if the intergration type has that type.
+func (i *IntegrationEnv) HasType(t string) bool {
+	_, ok := i.types[t]
+	return ok
+}
+
+// GetTypeSource returns the source file that determined that type.
+func (i *IntegrationEnv) GetTypeSource(t string) (string, bool) {
+	source, ok := i.types[t]
+	return source, ok
+}
+
+func haveIntegTestEnvRequirements(testEnv *IntegrationEnv) error {
+	if testEnv.HasType(TestingEnvDocker) || testEnv.HasType(TestingEnvKubernetes) {
```

Comment:
I see many `if testEnv.HasType(...)`, this will make problematic to add another provider (what will surely happen with terraform, or with future cloud-based kubernetes providers).
We could make `testEnv` to contain a list of "types" or providers, and each type has its way of check if it should be skipped, check their requirements, to execute and to cleanup. This way we can iterate on them and execute each operation instead of adding checks for each type.

Reply:
See latest commit for the solution to this problem.

A new runner can be easily registered with `RegisterIntegrationTester`.

I would like to add a scenarios runner that extends on this even more, that would allow things like inside of the kubernetes module:

```
module/kubernetes/integration_scenarios/aks/terraform.tf
module/kubernetes/integration_scenarios/aks/kubernetes.yml
module/kubernetes/integration_scenarios/eks/terraform.tf
module/kubernetes/integration_scenarios/eks/kubernetes.yml
module/kubernetes/integration_scenarios/gke/terraform.tf
module/kubernetes/integration_scenarios/gke/kubernetes.yml
```

Then the scenario runner would run the tests in each scenario `aks`, `eks`, `gke`. First using a `TerraformTestSetupStep` followed by the `KubernetesIntegrationTester`.

Execution order would be:

```
ScenarioIntegrationTester.Test
  - AWS
     - Terraform Step
     - Kubernetes Tester
  - EKS
     - Terraform Step
     - Kubernetes Tester
  - GKE
     - Terraform Step
     - Kubernetes Tester
```

Everything in the PR is designed to head down that path, looking forward to working on future branches to enable that work.